### PR TITLE
refactor(EntityService): Introduce a storage-independent in-memory entity aspect model

### PR DESCRIPTION
**Pull request by justinas-marozas**
## Why we want this change

The main goal of this PR is to reduce code duplication between `EntityService` implementations. The only expected differences between Cassandra and ebean implementations were concrete types representing an entity aspect. As the implementations are quite big, it's difficult to avoid unintentional divergences (some were already there).

We **don't** want to change DataHub behavior.

## What are the changes

* `EntityAspect` type was introduced to be used in public method signatures of `AspectDao` implementations instead of `EbeanAspectV2` and `CassandraAspect`;
* `EntityAspectIdentifier` type was introduced to abstract `EbeanAspectV2.PrimaryKey` away from public `AspectDao` method signatures;
* `AspectDao` interface was expanded to include everything `EntityService` implementations ask for;
* `EbeanEntityService`, `CassandraEntityService` and `EntityService` were merged. There's just `EntityService` now;
* `EbeanTimelineService` is now `TimelineServiceImpl` as `AspectDao` no longer exposes storage-specific types;
* `EbeanTimelineServiceTest` is now `TimelineServiceTest` and is arranged in the same manner as `EntityServiceTest` - an abstract class with all the tests accompanied by two implementations that set them up to run against their respective storage backends;
* `IngestDataPlatformInstancesStep` no longer depends on `EbeanServer`. Instead, it now depends on a new interface `AspectMigrationsDao` so it could work with Cassandra backend as well;
* A general purge of direct references to `EbeanServer`, `EbeanAspectDao`, `EbeanAspectV2` etc.

## What's not part of the changes

* `RetentionService` is still split into Cassandra and ebean implementations. More on that later;
* I left `com.linkedin.datahub.upgrade` types alone as these upgrade tasks are historic and storage specific. I'd also struggle to prove I hadn't broken them if I made significant changes.

### RetentionService

I meant to merge `RetentionService` as well but it'd proven complicated to get a clean split between "service" and "dao" concerns. With the way ebean implementation combines filters for version-based and time-based retentions into a single DB command, I couldn't think of a way to split it in a way where "service" knows nothing of ebean and "dao" knows nothing of retention. Running version and time based retentions in separate DB queries is also an option but I felt like I didn't have the time to test perf implications of that change.

In the end I left it alone as it neatly captures everything "retention" and my attempts to pull it apart took that away without giving anything in return.

### AspectMigrationsDao

The separation of `AspectDao` and `AspectMigrationsDao` is a bit of a mess as a single storage-specific type implements both of them, but I felt it's dangerous to mix normal `AspectDao` methods with the kinds of queries `IngestDataPlatformInstancesStep` makes (count _all_ entities etc).

I hope the name of this interface is appropriate for what `IngestDataPlatformInstancesStep` is doing.

### EntityAspectIdentifier

I'm not sure if this class is actually needed but some public `EbeanAspectDao` methods were accepting and returning `EbeanAspectV2.PrimaryKey`. I kinda like it because consuming code knows not to expect the rest of the `EntityAspect` attributes to have values, but getting rid of it wouldn't functionally make a difference.

I also think that `EntityAspectKey` would be a better name for what it is, but it could be confusing as "key aspect" is already a thing.

### CODE REVIEW

File: metadata-io/src/test/java/com/linkedin/metadata/entity/EntityServiceTest.java [hunk 1]
```diff
@@ -286,7 +313,7 @@ public void testIngestGetEntitiesV2() throws Exception {
 
         // Entity 2
         EntityResponse readEntityResponse2 = readEntities.get(entityUrn2);
-        assertEquals(2, readEntityResponse2.getAspects().size()); // Key + Info aspect.
+        assertEquals(readEntityResponse2.getAspects().size(), 2); // Key + Info aspect.
```
Comment (RyanHolstien) [inline metadata-io/src/test/java/com/linkedin/metadata/entity/EntityServiceTest.java:316]: Thank you! 😄

File: metadata-io/src/main/java/com/linkedin/metadata/timeline/TimelineServiceImpl.java [hunk 2]
```diff
@@ -166,10 +171,10 @@ public List<ChangeTransaction> getTimeline(@Nonnull final Urn urn, @Nonnull fina
         .filter(aspectSpec -> !aspectSpec.isTimeseries())
         .map(AspectSpec::getName)
         .collect(Collectors.toSet());
-    List<EbeanAspectV2> aspectsInRange = this._entityDao.getAspectsInRange(urn, fullAspectNames, startTimeMillis, endTimeMillis);
+    List<EntityAspect> aspectsInRange = this._aspectDao.getAspectsInRange(urn, fullAspectNames, startTimeMillis, endTimeMillis);
```
Comment (RyanHolstien) [inline metadata-io/src/main/java/com/linkedin/metadata/timeline/TimelineServiceImpl.java:174]: I think we should be able to push these DAO specific methods into an EntityService layer wrapper, what do you think? This would reduce the need for the split tests. Was this avoided due to level of effort or just trying to limit scope creep for changes?

**Review by RyanHolstien — APPROVED** [review]
This is awesome!! Thanks so much for this contribution @justinas-marozas 😄 Just some minor comments & questions

File: metadata-io/src/main/java/com/linkedin/metadata/timeline/TimelineServiceImpl.java (hunk 2 above)
Comment (justinas-marozas) [reply metadata-io/src/main/java/com/linkedin/metadata/timeline/TimelineServiceImpl.java:174]: I _would_ like to kill the need for TimelineService to have split tests. My mind was never on it because the diff of introducing EntityAspect was already spiralling out of control. 

Also, public methods in EntityService tend to wrap everything in RecordTemplates and if TimelineService used EntityService instead of the DAO - I assumed it (and the differs) would have to be rewired to use them instead of EntityAspect. I didn't feel like finding out how many changes that would be 😄

**Review by shirshanka — APPROVED** [review]
LGTM!
